src/vcpkg.cpp

```diff
@@ -41,6 +41,53 @@ static void invalid_command(const std::string& cmd)
     Checks::exit_fail(VCPKG_LINE_INFO);
 }
 
+static bool detect_container(vcpkg::Filesystem& fs)
+{
+    (void)fs;
+#if defined(_WIN32)
+    if (test_registry_key(HKEY_LOCAL_MACHINE, R"(SYSTEM\CurrentControlSet\Services\cexecsvc)"))
+    {
+        Debug::println("Detected Container Execution Service");
+        return true;
+    }
+
+    auto username = get_username();
+    if (!wcscmp(username.data(), L"ContainerUser") || !wcscmp(username.data(), L"ContainerAdministrator"))
+    {
+        Debug::println("Detected container username");
+        return true;
+    }
+#elif defined(__linux__)
+    if (fs.exists("/.dockerenv", IgnoreErrors{}))
+    {
+        Debug::println("Detected /.dockerenv file");
+        return true;
+    }
+
+    // check /proc/1/cgroup, if we're running in a container then the control group for each hierarchy will be:
+    //   /docker/<containerid>, or
+    //   /lxc/<containerid>
+    //
+    // Example of /proc/1/cgroup contents:
+    // 2:memory:/docker/66a5f8000f3f2e2a19c3f7d60d870064d26996bdfe77e40df7e3fc955b811d14
+    // 1:name=systemd:/docker/66a5f8000f3f2e2a19c3f7d60d870064d26996bdfe77e40df7e3fc955b811d14
+    // 0::/docker/66a5f8000f3f2e2a19c3f7d60d870064d26996bdfe77e40df7e3fc955b811d14
+    auto cgroup_contents = fs.read_contents("/proc/1/cgroup", IgnoreErrors{});
+    for (auto&& line : Strings::split(cgroup_contents, '\n'))
+    {
+        auto idx = line.rfind(':');
+        if (idx == std::string::npos) continue;
+        auto&& cgroup = line.substr(idx);
+        if (Strings::starts_with(cgroup, ":/docker/") || Strings::starts_with(cgroup, ":/lxc/"))
+        {
+            Debug::println("Detected docker in cgroup");
+            return true;
+        }
+    }
```

Comment:
Also, this block looks unit testable.

```suggestion
    for (auto&& line : Strings::split(cgroup_contents, '\n'))
    {
        auto idx = line.rfind(':');
        if (idx == std::string::npos) continue;
        auto cgroup = StringView{line}.substr(idx);
        if (Strings::starts_with(cgroup, ":/docker/") || Strings::starts_with(cgroup, ":/lxc/"))
        {
            Debug::println("Detected docker in cgroup");
            return true;
        }
    }
```